src/lib/operators/abstract_operator.cpp

```diff
@@ -24,20 +24,39 @@ namespace opossum {
 AbstractOperator::AbstractOperator(const OperatorType type, const std::shared_ptr<const AbstractOperator>& left,
                                    const std::shared_ptr<const AbstractOperator>& right,
                                    std::unique_ptr<AbstractOperatorPerformanceData> init_performance_data)
-    : performance_data(std::move(init_performance_data)), _type(type), _left_input(left), _right_input(right) {}
+    : performance_data(std::move(init_performance_data)), _type(type), _left_input(left), _right_input(right) {
+  // Tell input operators that we want to consume their output
+  if (_left_input) mutable_left_input()->register_consumer();
+  if (_right_input) mutable_right_input()->register_consumer();
+}
+
+AbstractOperator::~AbstractOperator() {
+  if constexpr (HYRISE_DEBUG && !HYRISE_TEST) {
+    bool left_has_executed = !_left_input ? false : _left_input->performance_data->executed;
+    bool right_has_executed = !_right_input ? false : _right_input->performance_data->executed;
+    Assert(_requested_for_execution || !(left_has_executed && right_has_executed),
+           "Operator has never been triggered for execution.");
+  }
```

Comment:
There are several tests that do not execute operators, but check other things like the output of `description()` for example.

Reply:
Ok, but why isn't the 'has_executed'-check sufficient?

Reply:
Because for some tests `execute()` is never called.

Reply:
Two thoughts on that:

* Why do we need `_requested_for_execution`? Why can't we use `this->performance_data->executed`?
* The `has_executed` logic is somewhat convoluted, starting with the negated condition on the ternary operator. In my understanding, we want the following behavior: If one of the input operators has been executed, this operator has to be executed before it is destructed, too. Based on that, the only reason I can see this assertion failing is if we build a PQP and only execute some operators in it. Is that the case for all failing tests? Could you give an example?

Reply:
> Ok, but why isn't the 'has_executed'-check sufficient?

Oh, you might have asked because of `_requested_for_execution`. Sometimes operators fail and `has_executed` does not get set.

Reply:
Do you mean "fail" in the context of MVCC or do you mean actual failures?

Reply:
In the context of MVCC.

Reply:
Are you talking about

```
if (transaction_context->aborted()) {
  return;
}
```?

In that case, maybe we should add the `transaction_context->aborted()` condition to the assert that this discussion is about?

Reply:
Good idea. But, since we only hold a weak reference to the TA-context, the TA-context is not always available.

Reply:
I added a hack condition to the Assert to fix the tests.

Reply:
> Good idea. But, since we only hold a weak reference to the TA-context, the latter is not always available.

Did this actually pop up as an issue?

Reply:
Yes, for some tests it did.